(Introduce the project)
(Introduce the dataset)
(Analysis focus: Type of dogs with highest ratings, dog stage with highest ratings, Type of dogs with the highest favourite_count, Type of dog with highest retweet_count, dog stage and favourite_count and retweet_count)

## Gather

In [37]:
# KEEP
import json
import numpy as np
import os
import pandas as pd
import requests
import tweepy
from timeit import default_timer as timer
from tweepy import OAuthHandler

In [2]:
# KEEP
# Load twitter archive file
twitter_archive_enhanced = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# KEEP
# Download image-predictions file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open('image_predictions.tsv', mode='wb') as file:
    file.write(response.content)

In [4]:
# KEEP
# Import image predictions file
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')

In [5]:
# KEEP
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

"\nconsumer_key = 'HIDDEN'\nconsumer_secret = 'HIDDEN'\naccess_token = 'HIDDEN'\naccess_secret = 'HIDDEN'\n\nauth = OAuthHandler(consumer_key, consumer_secret)\nauth.set_access_token(access_token, access_secret)\n\napi = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)\n"

In [6]:
# KEEP
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = twitter_archive_enhanced.tweet_id.values
len(tweet_ids)

2356

In [38]:
# KEEP
# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
if not os.path.exists('tweet_json.txt'):
    count = 0
    fails_dict = {}
    start = timer()
    # Save each tweet's returned JSON as a new line in a .txt file
    with open('tweet_json.txt', 'w') as outfile:
        # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
        for tweet_id in tweet_ids:
            count += 1
            print(str(count) + ": " + str(tweet_id))
            try:
                tweet = api.get_status(tweet_id, tweet_mode='extended')
                print("Success")
                json.dump(tweet._json, outfile)
                outfile.write('\n')
            except tweepy.TweepError as e:
                print("Fail")
                fails_dict[tweet_id] = e
                pass
    end = timer()
    print(end - start)
    print(fails_dict)


In [8]:
# DELETE
# Temp - save fails_dict to file for working with
'''
import csv
fails_list = []
for key in fails_dict.keys():
    fails_list.append(key)
fails_list
with open('fails_dict.csv', 'w') as outfile:
    wr = csv.writer(outfile, dialect='excel')
    wr.writerow(fails_list)
'''

"\nimport csv\nfails_list = []\nfor key in fails_dict.keys():\n    fails_list.append(key)\nfails_list\nwith open('fails_dict.csv', 'w') as outfile:\n    wr = csv.writer(outfile, dialect='excel')\n    wr.writerow(fails_list)\n"

In [9]:
# DELETE
# Working - uncomment out if necessary
'''
failures = []
with open('fails_dict.csv', 'r') as f:
    reader = csv.reader(f)
    for line in reader:
        failures.append(line)
'''

"\nfailures = []\nwith open('fails_dict.csv', 'r') as f:\n    reader = csv.reader(f)\n    for line in reader:\n        failures.append(line)\n"

In [10]:
# KEEP
# Load extended tweet data
tweets = []
with open('tweet_json.txt') as json_file:
    for line in json_file:
        tweets.append(json.loads(line))

# Place extended tweet data into a dataframe
extended_data = pd.DataFrame(tweets)

## Assess

(Explore each using .info)

In [11]:
# KEEP
# View column types
twitter_archive_enhanced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [12]:
# DETERMINE IF TO KEEP
# Visually assess data in archive
twitter_archive_enhanced.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1048  743253157753532416                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
538   806620845233815552                    NaN                  NaN   
1681  681981167097122816                    NaN                  NaN   
2117  670421925039075328                    NaN                  NaN   

                      timestamp  \
1048  2016-06-16 01:25:36 +0000   
1843  2015-12-13 01:41:41 +0000   
538   2016-12-07 22:06:10 +0000   
1681  2015-12-29 23:32:35 +0000   
2117  2015-11-28 02:00:17 +0000   

                                                 source  \
1048  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   
538   <a href="http://twitter.com/download/iphone" r...   
1681  <a href="http://twitter.com/download/iphone" r...   
2117  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1048  This is Kilo. He cannot reach the snackum. Nif...                  NaN   
1843  Here we have an entire platoon of puppers. Tot...                  NaN   
538   RT @dog_rates: Idk why this keeps happening. W...         7.815247e+17   
1681  This is Jimothy. He's a Trinidad Poliwhirl. Fa...                  NaN   
2117           Meet Herb. 12/10 https://t.co/tLRyYvCci3                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1048                       NaN                        NaN   
1843                       NaN                        NaN   
538               4.196984e+09  2016-09-29 16:03:01 +0000   
1681                       NaN                        NaN   
2117                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1048  https://twitter.com/dog_rates/status/743253157...                10   
1843  https://twitter.com/dog_rates/status/675853064...                88   
538   https://twitter.com/dog_rates/status/781524693...                12   
1681  https://twitter.com/dog_rates/status/681981167...                12   
2117  https://twitter.com/dog_rates/status/670421925...                12   

      rating_denominator     name doggo floofer pupper  puppo  
1048                  10     Kilo  None    None   None  puppo  
1843                  80     None  None    None   None   None  
538                   10     None  None    None   None   None  
1681                  10  Jimothy  None    None   None   None  
2117                  10     Herb  None    None   None   None

In [14]:
# KEEP
# Check for duplicated tweets
len(twitter_archive_enhanced[twitter_archive_enhanced.tweet_id.duplicated()])

0

In [15]:
# KEEP
# Check values in rating_numerator
twitter_archive_enhanced.rating_numerator.value_counts().sort_index().head()

0     2
1     9
2     9
3    19
4    17
Name: rating_numerator, dtype: int64

A visual assessment of the data file shows rating_numerator in the text column of negative value and partial value (e.g. 3.5). These have not been identified correctly as they are not present in the rating_numerator column.

In [16]:
# KEEP
# Check values in rating_denominator
twitter_archive_enhanced.rating_denominator.value_counts().sort_index().head()

0        1
2        1
7        1
10    2333
11       3
Name: rating_denominator, dtype: int64

In [17]:
# KEEP
# Check for inclusion of retweets
sum(~twitter_archive_enhanced.retweeted_status_id.isnull())

181

There are 181 tweets that are retweets and need to be removed.

In [18]:
# KEEP
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [19]:
# DETERMINE IF TO KEEP
image_predictions.tail(15)

tweet_id                                          jpg_url  \
2060  889531135344209921  https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072  https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689  https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881  https://pbs.twimg.com/media/DFl99B1WsAITKsg.jpg   
2064  890006608113172480  https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg   
2065  890240255349198849  https://pbs.twimg.com/media/DFrEyVuW0AAO3t9.jpg   
2066  890609185150312448  https://pbs.twimg.com/media/DFwUU__XcAEpyXI.jpg   
2067  890729181411237888  https://pbs.twimg.com/media/DFyBahAVwAAhUTd.jpg   
2068  890971913173991426  https://pbs.twimg.com/media/DF1eOmZXUAALUcq.jpg   
2069  891087950875897856  https://pbs.twimg.com/media/DF3HwyEWsAABqE6.jpg   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                        p1   p1_conf  p1_dog                  p2  \
2060        1          golden_retriever  0.953442    True  Labrador_retriever   
2061        1            French_bulldog  0.991650    True               boxer   
2062        1                  Pembroke  0.966327    True            Cardigan   
2063        1            French_bulldog  0.377417    True  Labrador_retriever   
2064        1                   Samoyed  0.957979    True          Pomeranian   
2065        1                  Pembroke  0.511319    True            Cardigan   
2066        1             Irish_terrier  0.487574    True        Irish_setter   
2067        2                Pomeranian  0.566142    True          Eskimo_dog   
2068        1               Appenzeller  0.341703    True       Border_collie   
2069        1  Chesapeake_Bay_retriever  0.425595    True       Irish_terrier   
2070        2                    basset  0.555712    True    English_springer   
2071        1               paper_towel  0.170278   False  Labrador_retriever   
2072        1                 Chihuahua  0.716012    True            malamute   
2073        1                 Chihuahua  0.323581    True            Pekinese   
2074        1                    orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
2060  0.013834    True                      redbone  0.007958    True  
2061  0.002129    True    Staffordshire_bullterrier  0.001498    True  
2062  0.027356    True                      basenji  0.004633    True  
2063  0.151317    True                       muzzle  0.082981   False  
2064  0.013884    True                         chow  0.008167    True  
2065  0.451038    True                    Chihuahua  0.029248    True  
2066  0.193054    True     Chesapeake_Bay_retriever  0.118184    True  
2067  0.178406    True                     Pembroke  0.076507    True  
2068  0.199287    True                    ice_lolly  0.193548   False  
2069  0.116317    True              Indian_elephant  0.076902   False  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False

In [20]:
# KEEP
extended_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2339 entries, 0 to 2338
Data columns (total 32 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2339 non-null object
display_text_range               2339 non-null object
entities                         2339 non-null object
extended_entities                2065 non-null object
favorite_count                   2339 non-null int64
favorited                        2339 non-null bool
full_text                        2339 non-null object
geo                              0 non-null object
id                               2339 non-null int64
id_str                           2339 non-null object
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null object
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 non-null obj

In [21]:
# DETERMINE IF TO KEEP
extended_data.head(5)

contributors coordinates                      created_at display_text_range  \
0         None        None  Tue Aug 01 16:23:56 +0000 2017            [0, 85]   
1         None        None  Tue Aug 01 00:17:27 +0000 2017           [0, 138]   
2         None        None  Mon Jul 31 00:18:03 +0000 2017           [0, 121]   
3         None        None  Sun Jul 30 15:58:51 +0000 2017            [0, 79]   
4         None        None  Sat Jul 29 16:00:24 +0000 2017           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           37783   
1  {'media': [{'id': 892177413194625024, 'id_str'...           32453   
2  {'media': [{'id': 891815175371796480, 'id_str'...           24437   
3  {'media': [{'id': 891689552724799489, 'id_str'...           41113   
4  {'media': [{'id': 891327551943041024, 'id_str'...           39326   

   favorited                                          full_text   geo  ...  \
0      False  This is Phineas. He's a mystical boy. Only eve...  None  ...   
1      False  This is Tilly. She's just checking pup on you....  None  ...   
2      False  This is Archie. He is a rare Norwegian Pouncin...  None  ...   
3      False  This is Darla. She commenced a snooze mid meal...  None  ...   
4      False  This is Franklin. He would like you to stop ca...  None  ...   

   quoted_status quoted_status_id quoted_status_id_str  \
0            NaN              NaN                  NaN   
1            NaN              NaN                  NaN   
2            NaN              NaN                  NaN   
3            NaN              NaN                  NaN   
4            NaN              NaN                  NaN   

   quoted_status_permalink retweet_count  retweeted retweeted_status  \
0                      NaN          8233      False              NaN   
1                      NaN          6084      False              NaN   
2                      NaN          4027      False              NaN   
3                      NaN          8384      False              NaN   
4                      NaN          9089      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

In [31]:
# KEEP
# Find number of retweets in extended_data
len(extended_data[~extended_data['retweeted_status'].isnull()])

167

In [35]:
extended_data.retweet_count.describe()

count     2339.000000
mean      2903.120992
std       4894.869626
min          0.000000
25%        583.500000
50%       1354.000000
75%       3383.500000
max      82968.000000
Name: retweet_count, dtype: float64

In [33]:
# KEEP
# Check for duplication between twitter_archive_enhanced and image_predictions
all_columns = pd.Series(list(twitter_archive_enhanced) + list(image_predictions))
all_columns[all_columns.duplicated()]

17    tweet_id
dtype: object

Only the tweet_id column was duplicated, which we can use as the link between to two tables.

#### Quality
##### `twitter_archive_enhanced` table
- tweet_id is an int (should be str)
- rating_numerator and rating_denominator columns are ints (should be floats)
- rating_numerator and rating_denominator columns incorrect (missing decimal places, negative numbers recorded as positive numbers)
- presence of retweets (retweeted_status_id is non-null) in twitter_archive_enhanced



##### `image_predictions` table
- tweet_id is an int (should be str)
- dog names in px column inconsistent - upper and lowercase names, inclusion of underscores etc.
- predictions include non-dog objects which should be removed prior to analysis

##### `extended_data` table
- presence of retweets (retweeted_status is non-null) in extended_data
- x
- x

#### Tidiness
- twitter_archive_enhanced contains multiple columns for the dog stage when should be one column with the dog stage as the variable
- image_predictions table has multiple predictions per row
- extended_info should be part of archive (keeping favorite_count and retweet_count)

## Clean

(make copies of datasources)

In [23]:
# KEEP
archive_clean = archive.copy()
image_predictions_clean = image_predictions.copy()
extended_data_clean = extended_data.copy()

NameError: name 'archive' is not defined

### (Problem 1)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 2)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 3)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 4)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 5)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 6)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 7)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 8)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 9)

(Problem: ...)

##### Define
x

##### Code

##### Test

### (Problem 10)

(Problem: ...)

##### Define
x

##### Code

##### Test

## Storage

(store each table as a csv file with main file called tiwtter_archive_master.csv)

## Analysis

(make at least 3 insights)